src/main/scala/net/snowflake/spark/snowflake/ServerConnection.scala

```diff
@@ -274,6 +290,36 @@ private[snowflake] object ServerConnection {
     jdbcConnectionCount.incrementAndGet()
     conn
   }
+
+  object providedConnections {
+    // use provided JDBC connection
+    private val providedConn = mutable.Map[String, Connection]()
+    def register(conn: Connection): String = {
+      val sessionId: String = conn.asInstanceOf[SnowflakeConnectionV1].getSessionID
+      providedConn.put(sessionId, conn)
+      sessionId
+    }
+    def hasConnectionID(id: String): Boolean =
+      providedConn.contains(id)
+
+    def getConnection(id: String): Option[Connection] = this.synchronized {
+      val result = providedConn.get(id)
+      if (result.isDefined) {
+        providedConn.remove(id)
```

Comment:
Remove the cached the connection from staging area after get. Connector will use the regular shared connection feature to cache the provided connection.